python_ta/contracts/__init__.py

```diff
@@ -1,40 +1,70 @@
+    """Exception raised when a condition is invalid
     """
 
+
 @wrapt.decorator
-def check_contracts(wrapped, _instance, args, kwargs):
+def check_contracts(wrapped, instance, args, kwargs):
+    if instance is None and inspect.isclass(wrapped):
+        return add_class_contracts(wrapped, args, kwargs)
+    return check_function_contracts(wrapped, instance, args, kwargs)
+
+
+def add_class_contracts(wrapped, args, kwargs):
+    rep_invariants = parse_conditions(
+        wrapped.__doc__ or '', parse_token="Representation Invariant")
+    setattr(wrapped, "__representation_invariants__", rep_invariants)
+    return wrapped(*args, **kwargs)
+
+
+def check_function_contracts(wrapped, instance, args, kwargs):
     params = wrapped.__code__.co_varnames[:wrapped.__code__.co_argcount]
     annotations = typing.get_type_hints(wrapped)
-    for arg, param in zip(args, params):
+
+    args_with_self = (instance,) + args if instance else args
+
+    for arg, param in zip(args_with_self, params):
         if param in annotations:
             assert check_type_annotation(annotations[param], arg),\
                 f'Argument {repr(arg)} did not match type annotation for parameter "{param}: {annotations[param]}"'
 
-    preconditions = parse_preconditions(wrapped.__doc__ or '')
-    function_locals = dict(zip(params, args))
-    for precondition in preconditions:
-        try:
-            check = eval(precondition, wrapped.__globals__, function_locals)
-        except:
-            # TODO: Decide what to do here, e.g. "Invalid precondition"
-            raise InvalidPrecondition(
-                f'Error evaluating precondition: {precondition}')
-        else:
-            assert check,\
-                f'Precondition "{precondition}" violated for arguments {function_locals}'
+    preconditions = parse_conditions(wrapped.__doc__ or '')
+
+    function_locals = dict(zip(params, args_with_self))
+    validate_conditions(wrapped, function_locals, preconditions)
+
     r = wrapped(*args, **kwargs)
+
+    if instance and hasattr(type(instance), "__representation_invariants__"):
+        validate_conditions(wrapped, function_locals,
+                            type(instance).__representation_invariants__)
+
     if 'return' in annotations:
         return_type = annotations['return']
         assert check_type_annotation(return_type, r),\
             f'Return value {r} does not match annotated return type {return_type.__name__}'
     return r
 
+
+def validate_conditions(wrapped, function_locals, conditions):
+    for condition in conditions:
+        try:
+            check = eval(condition, wrapped.__globals__, function_locals)
+        except:
+            # TODO: Decide what to do here, e.g. "Invalid condition"
+            raise InvalidCondition(
+                f'Error evaluating condition: {condition}')
+        else:
+            assert check,\
+                f'Condition "{condition}" violated for arguments {function_locals}'
```

Comment:
Okay this is interesting. I'd like to have separate error messages for the violation of a precondition vs. representation invariants. (Indeed, `function_locals` is too broad when it's only the `self` argument being checked for a representation invariant.)

I'll let you decide how you want to handle this: either with another parameter, or using a `try`-`except` to handle the `AssertionError` in context, etc.

Reply:
I now have a separate function to check invariants hence there is a different error message.